lib/plugins/threaded-signing/plugin.c

```diff
@@ -254,30 +281,34 @@ threaded_openssl_get_signature(sv_threaded_plugin_t *self,
   SignedVideoReturnCode status = SV_OK;
 
   g_mutex_lock(&self->mutex);
-  if (self->plugin_state == THREADED_SIGNING_HAS_SIGNATURE && self->signature_info) {
-    if (self->signature_info->signature_size > max_signature_size) {
+  if (self->output_buffer[0].signature) {
+    if (self->output_buffer[0].size > max_signature_size) {
       // If there is no room to copy the signature, report zero size.
       *written_signature_size = 0;
     } else {
-      memcpy(signature, self->signature_info->signature, self->signature_info->signature_size);
-      *written_signature_size = self->signature_info->signature_size;
+      // Get the oldest signature
+      memcpy(signature, self->output_buffer[0].signature, self->output_buffer[0].size);
+      *written_signature_size = self->output_buffer[0].size;
+      // Free |signature| and move buffer
+      const int output_buffer_end = self->output_buffer_idx;
+      free(self->output_buffer[0].signature);
+      self->output_buffer[0].signature = NULL;
+      for (int j = 0; j < output_buffer_end - 1; j++) {
+        self->output_buffer[j].signature = self->output_buffer[j + 1].signature;
+        self->output_buffer[j].size = self->output_buffer[j + 1].size;
+      }
+      self->output_buffer[output_buffer_end - 1].signature = NULL;
+      self->output_buffer[output_buffer_end - 1].size = 0;
+      self->output_buffer_idx -= 1;
     }
     // Change state and mark as copied.
-    self->plugin_state = THREADED_SIGNING_WAITS_FOR_HASH_TO_SIGN;
     has_copied_signature = true;
   } else if (self->plugin_state == THREADED_SIGNING_ERROR) {
```

Comment:
This is the only place where 'plugin_state' is used. Right now it does not work for mainly two reasons. 1) it may be set for a different buffer element and, 2) you still copy to the output buffer even if the signing fails.